### ISSUE TITLE
`Simulation.minimizeEnergy` not working for me

### ISSUE THREAD

**Issue by swails** [pandegroup/openmm#1229]
This is a preliminary report, but I've been finding that applying the OpenMM minimizer to my system is actually making things *worse* than they were before.

Not only is it not resolving bad contacts, it is actually *creating* some where none were before.  Since it doesn't print a progress log, it's impossible for me to track what's happening at each iteration without hacking the LBFGS code to do so.

Here is my general workflow -- I create and solvate a system using tleap for use with the Amoeba force field (I have a fix in the works to let `Modeller` work with a `ForceField` that doesn't assign a `NonbondedForce`, but it doesn't currently work).  I then take this PDB file, generate a `System` from that, and then minimize it using the following script:

```Python
#!/usr/bin/env python
from __future__ import print_function, division

import parmed as pmd

import simtk.openmm as mm
import simtk.openmm.app as app

print('Loading the system and PDB file...')
system = pmd.load_file('system.xml')
parm = pmd.load_file('2igd_solvated.pdb')

print('Creating the Simulation...')
sim = app.Simulation(parm.topology, system, mm.VerletIntegrator(0.001),
                     platform=mm.Platform.getPlatformByName('CUDA'),
                     platformProperties=dict(CudaPrecision='mixed'))

print('Minimizing...')
sim.context.setPositions(parm.positions)
sim.minimizeEnergy(1.0)

# Now write a serialized state that has coordinates
print('Finished. Writing serialized XML restart file...')
with open('2igd_solvated.min.xml', 'w') as f:
    f.write(
            mm.XmlSerializer.serialize(
                sim.context.getState(getPositions=True, getVelocities=True,
                                     getForces=True, getEnergy=True,
                                     enforcePeriodicBox=True)
            )
    )

print('Done!')
```

I check my structure for bad contacts using the `checkoverlaps` command in cpptraj which scans through all interatomic distances and reports any that fall below 0.8 A.  If I use ParmEd to take the minimized state XML file and create a new PDB (since cpptraj can't read serialized states), I can feed it through and get the close contacts.  The list I get following minimization is *longer* than the list I get for the original PDB.

Sure enough, as soon as I try to run a simulation with this system, it runs off to NaNdyland and never comes back :(.

The simulation gets farther when I don't minimize at all (but this is rarely an option).  The next thing I tried was hijacking the `scipy` LBFGS optimizer and using that instead via this script:

```Python
#!/usr/bin/env python

import parmed as pmd
import parmed.unit as u
import simtk.openmm as mm
from scipy import optimize

parm = pmd.load_file('2igd_solvated.pdb')
system = pmd.load_file('system.xml')

con = mm.Context(system, mm.VerletIntegrator(0.001),
                 mm.Platform.getPlatformByName('CUDA'))
con.setPositions(parm.positions)

def target_function(xyz):
    con.setPositions(xyz.reshape(-1,3))
    state = con.getState(getEnergy=True, getForces=True)
    frc = state.getForces(asNumpy=True)
    ene = state.getPotentialEnergy().value_in_unit(u.kilojoule_per_mole)
    frc = frc.value_in_unit(u.kilojoule_per_mole/u.nanometer)
    return ene, -frc.flatten()

results = optimize.minimize(target_function, parm.coordinates/10, method='L-BFGS-B',
                            jac=True, options=dict(maxiter=10000, disp=True, gtol=0.1))

parm.coordinates = results.x * 10

parm.save('minimized_scipy.pdb')
```

The simulation I started from this PDB file is still running (although I have less than 1 total ns so far).  So something is definitely fishy with the OpenMM minimizer -- at least for me with the Amoeba simulation.  The scipy minimizer seems to do a nice job.  Analyzing with `checkoverlaps` in cpptraj after it's finished results in 0 close contacts (compared to a couple dozen with the OpenMM minimizer).

A couple observations here:

- The starting structure does *not* have many problems.  In fact, a simulation *can* run without minimizing (although the *second* simulation seems to have problems).
- I picked the LBFGS algo in scipy, too, so the algorithms should be fairly similar (although not the same).

I packed up the necessary files here: https://www.dropbox.com/s/qdjpqx8ao1gotjj/amoeba_minimizer.tgz?dl=0

**Comment by jchodera** [comment]
We have seen similar behavior as well, though I suspected this was due to the fact that the minimizer does not really enforce constraints, but instead adds a harmonic penalty term.

Does anything change if you start by enforcing the constraints and then minimize?

**Comment by jchodera** [comment]
I have a preliminary implementation of the FIRE minimizer in a branch of openmmtools if you want to give that a try.

**Comment by swails** [comment]
This system has no constraints. 

--
Jason M. Swails
BioMaPS,
Rutgers University
Postdoctoral Researcher

**Comment by jchodera** [comment]
Ah, right---AMOEBA.

**Comment by rmcgibbo** [comment]
Do you see a significantly different final energy between the OpenMM minimizer and scipy? I added instrumenting to `LocalEnergyMinimizer.cpp` ( https://github.com/pandegroup/openmm/compare/master...rmcgibbo:instrument-minimizer), and ran your code. [Here](https://gist.github.com/rmcgibbo/a5d2c7fd1b07fca21bceq) are the logs. They come out basically the same.

**Comment by rmcgibbo** [comment]
I do see a difference in the number of close contacts, but I'm tempted to blame this on the Amoeba FF, not the optimizer.

**Comment by swails** [comment]
The link you posted to the logs gives me 404...

**Comment by swails** [comment]
Thanks for the code @rmcgibbo!  I checked out your instrument-minimizer branch and I think I'm getting a little closer to the issue.

I agree the minimizer looks fine.  However, the final XML file that gets dumped *following* the minimization seems clearly wrong.  Here's the `minimize.py` script that I modified:

```Python
#!/usr/bin/env python
from __future__ import print_function, division

import parmed as pmd
import parmed.unit as u

import simtk.openmm as mm
import simtk.openmm.app as app

print('Loading the system and PDB file...')
system = pmd.load_file('system.xml')
parm = pmd.load_file('2igd_solvated.pdb')

print('Creating the Simulation...')
sim = app.Simulation(parm.topology, system, mm.VerletIntegrator(0.001),
                     platform=mm.Platform.getPlatformByName('CUDA'),
                     platformProperties=dict(CudaPrecision='mixed'))

print('Minimizing...')
sim.context.setPositions(parm.positions)
sim.minimizeEnergy(1.0, verbose=True)

# Now write a serialized state that has coordinates
print('Finished. Writing serialized XML restart file...')
with open('2igd_solvated.min.xml', 'w') as f:
    f.write(
            mm.XmlSerializer.serialize(
                sim.context.getState(getPositions=True, getVelocities=True,
                                     getForces=True, getEnergy=True,
                                     enforcePeriodicBox=True)
            )
    )

print('Loading the XML file and calculating energy')
sim.context.setState(
        mm.XmlSerializer.deserialize(open('2igd_solvated.min.xml').read())
)
state = sim.context.getState(getEnergy=True)
print('Potential energy is %.5f' %
        (state.getPotentialEnergy().value_in_unit_system(u.md_unit_system))
)

print('Done!')
```

So all I did here was dump a state XML file and then immediately read it back in.  When I recalculated the energy, I got a *huge* value.  Here is the tail of the output from the script (running in @rmcgibbo's branch that has minimization tracking).

```
At iterate  4246    f= -3.656453e+05    |g|= 1.896832e+02
At iterate  4247    f= -3.656454e+05    |g|= 2.957360e+02
At iterate  4248    f= -3.656454e+05    |g|= 3.824114e+02
At iterate  4249    f= -3.656455e+05    |g|= 1.808041e+02
At iterate  4250    f= -3.656455e+05    |g|= 1.457220e+02
Finished. Writing serialized XML restart file...
Loading the XML file and calculating energy
Potential energy is 213259567.67776
Done!
```

As you can see, the potential energy from the XML file loaded after the minimization completed is +2.13e+8 kJ/mol.  By contrast, the minimized energy, according to the reporter, is ~-3.656455e5 kJ/mol.  So something is clearly wrong -- either my expectation of what `context.getState` should do after `Simulation.minimizeEnergy` or something else.

The reason I didn't see this with the scipy minimizer is because it skipped over the `getState`/`setState` step altogether, just giving me an array of coordinates that I dumped to a PDB file for later use.

I also can't seem to restart my simulation with a dumped XML file following the scipy minimization, as the simulation blows up.  I suspect something similar happening there, although I'm not quite sure at this point where to start looking.

**Comment by swails** [comment]
Well now this is interesting.  I inserted the following code *right* before dumping the state XML file in the script:

```Python
print('Calculating energy...')
state = sim.context.getState(getEnergy=True)
print('Energy = %.5f' %
        (state.getPotentialEnergy().value_in_unit_system(u.md_unit_system))
)
```

This completely changed the output of the program:

```
At iterate  3942    f= -3.651441e+05    |g|= 1.800481e+02
At iterate  3943    f= -3.651442e+05    |g|= 1.964849e+02
At iterate  3944    f= -3.651442e+05    |g|= 5.234011e+02
At iterate  3945    f= -3.651443e+05    |g|= 1.766326e+02
At iterate  3946    f= -3.651443e+05    |g|= 1.305794e+02
Calculating energy...
Energy = -365144.33052
Finished. Writing serialized XML restart file...
Loading the XML file and calculating energy
Potential energy is -93903087.24604
Done!
```

As you can see, the energy I get in the initial energy calculation matches what the minimizer reported.  But dumping then loading the state XML file results in a completely whacked-out energy.  Something is wrong there.

**Comment by rmcgibbo** [comment]
Maybe something with enforcePeriodic in getState?

**Comment by swails** [comment]
> Maybe something with enforcePeriodic in getState?

Good call.  If I remove `enforcePeriodicBox`, the energy matches again:

```
At iterate    45    f= -3.195697e+05    |g|= 3.305029e+04
At iterate    46    f= -3.200699e+05    |g|= 2.857384e+04
At iterate    47    f= -3.204872e+05    |g|= 1.682629e+04
At iterate    48    f= -3.209656e+05    |g|= 1.448297e+04
Calculating energy...
Energy = -320965.60920
Calculating energy again...
Energy = -320965.60446
Loading the XML file and calculating energy
Potential energy is -320965.60760
Done!
```

**Comment by swails** [comment]
One thing I should point out is that this system is a truncated octahedron.

@peastman -- could it be that the `enforcePeriodicBox` functionality in `getState` wasn't updated for triclinic cells?

**Comment by swails** [comment]
Actually I think that's the key ingredient here -- I'm using the same workflow (saving a state.xml via `enforcePeriodicBox=True` and not seeing any of these problems, but that system is a crystal simulation with an orthorhombic unit cell.

I think `enforcePeriodicBox=True` is broken for non-orthogonal cells.  Which would certainly explain some of the recent struggles I've been having with some of my simulations blowing up in OpenMM when they ran just fine in Amber (I always use truncated octahedra when I can).

Unless @peastman beats me to it, I'll see if I can get a patch in tomorrow sometime for this problem.

**Comment by rmcgibbo** [comment]
Fixed by #1234.
